src/core/namespace.c

```diff
@@ -3155,6 +3157,113 @@ bool ns_type_supported(NamespaceType type) {
         return access(ns_proc, F_OK) == 0;
 }
 
+int refresh_extensions_in_namespace(
+                const PidRef *target,
+                const char *hierarchy_env,
+                const NamespaceParameters *p) {
+
+        const char *overlay_prefix = "/run/systemd/mount-rootfs";
+        _cleanup_(mount_list_done) MountList ml = {};
+        _cleanup_free_ char *extension_dir = NULL;
+        _cleanup_strv_free_ char **hierarchies = NULL;
+        pid_t pid;
+        int r;
+
+        assert(target);
+        assert(hierarchy_env);
+        assert(p);
+
+        log_debug("Refreshing extensions in-namespace for hierarchy '%s'", hierarchy_env);
+
+        extension_dir = strjoin(p->private_namespace_dir, "/unit-extensions");
+        if (!extension_dir)
+                return -ENOMEM;
+
+        r = parse_env_extension_hierarchies(&hierarchies, hierarchy_env);
+        if (r < 0)
+                return r;
+
+        r = append_extensions(
+                        &ml,
+                        overlay_prefix,
+                        p->private_namespace_dir,
+                        hierarchies,
+                        p->extension_images,
+                        p->n_extension_images,
+                        p->extension_directories);
+        if (r < 0)
+                return r;
+        if (ml.n_mounts == 0)
+                return 0;
+
+        (void) mkdir_p(overlay_prefix, 0755);
+
+        r = safe_fork("(sd-ns-refresh-exts)",
+                        FORK_DEATHSIG_SIGTERM | FORK_LOG | FORK_WAIT | FORK_NEW_MOUNTNS | FORK_MOUNTNS_SLAVE,
```

Comment:
I don't think FORK_MOUNTNS_SLAVE should be needed here?

Reply:
Without it, the mounts keep persisting in between calls, and keeping `FORK_MOUNTNS_SLAVE` nicely prevents this behaviour. Perhaps I could get rid of `propagate_path`/ `incoming_path` so that it wouldn't try to set up the propagation (but `bind_mount_in_namespace` will still want these two paths)?

Reply:
right I always forget what is the default for propagation, never mind